conans/client/cmd/uploader.py

```diff
@@ -150,6 +150,259 @@ def _collect_packages_to_upload(self, refs, confirm, remotes, all_packages, quer
         return refs_by_remote
 
 
+class _PackagePreparator(object):
+    def __init__(self, cache, remote_manager, hook_manager, output):
+        self._cache = cache
+        self._remote_manager = remote_manager
+        self._output = output
+        self._hook_manager = hook_manager
+
+    def prepare_recipe(self, ref, conanfile, remote, remotes, policy):
+        """ do a bunch of things that are necessary before actually executing the upload:
+        - retrieve exports_sources to complete the recipe if necessary
```

Comment:
Is it too ambitious to separate this step out even further. Recall that we have multiple people who would like to see a new command or flag added which can force conan to retrieve the exports_sources for debug purposes.

Reply:
Please note that this is only calling the ``retrieve_exports_sources(self._remote_manager, self._cache, conanfile, ref, remotes)``, which is only getting the ``export_source`` files, but not the actual complete source files necessary to build and debug, that will be the result of merging that with exported files, scm and any other action that the ``source()`` method is doing. So it seems the entry point for that functionality would be closer to the ``source.py`` files, and shouldn't be extracted from here.